### PR TITLE
Move Crawler to Dashboard & Refactor crawler

### PR DESCRIPTION
**Pull request by Andrewmatilde**
Use github.com/shurcooL/graphql instead of github.com/machinebox/graphql and give schema info to simpify writing query.

### CODE REVIEW

File: infrastructure/github/crawler/client/client.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+	client = graphql.NewClient(config.ServerUrl)
+	for i, s := range config.Authorization {
+		req := graphql.NewRequest(string(content))
+		req.Header.Set("authorization", "Bearer "+s)
+		requests[i] = req
+	}
+	clientIsOpen = true
+}
+
+type AuthUseOut struct {
```
Comment (mahjonp) [inline infrastructure/github/crawler/client/client.go:47]: Should be prefixed with Err

File: infrastructure/github/crawler/crawler.go [hunk 2]
```diff
@@ -0,0 +1,155 @@
+package crawler
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"dashboard/infrastructure/github/crawler/client"
```
Comment (mahjonp) [inline infrastructure/github/crawler/crawler.go:8]: I suggest renaming the module to the full name(github.com/PingCAP-Q/dashboard), it looks strange here.

File: infrastructure/github/crawler/crawler.go [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+	"github.com/pkg/math"
+)
+
+const maxGithubPageSize = 100
+
+type FetchOption struct {
+	Owner        string
+	RepoName     string
+	First        *int
+	IssueFilters *map[string]interface{}
```
Comment (mahjonp) [inline infrastructure/github/crawler/crawler.go:22]: Usually, it's better to use `map[string]interface{}`, no need to use a pointer IMO.

File: infrastructure/github/crawler/crawler.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+		"repo_name":       opt.RepoName,
+		"issueFilters":    opt.IssueFilters,
+		"issueAfter":      nil,
+		"commentAfter":    nil,
+		"tagAfter":        nil,
+		"IssuePageSize":   0,
+		"CommentPageSize": 0,
+		"tagPageSize":     0,
+	}
+	fmt.Printf("Ping count with %v\n", v)
```
Comment (mahjonp) [inline infrastructure/github/crawler/crawler.go:46]: Suggest using a structured logger for better use of the log monitor.

File: infrastructure/github/crawler/crawler.go [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+	fmt.Printf("Get data issue count: %d \n", totalCount)
+	for count := 0; count < totalCount; count += math.Min(totalCount-count, maxGithubPageSize) {
+		fmt.Printf("<Fetching issue data %d to %d\n", count, count+math.Min(totalCount-count, maxGithubPageSize))
+		v["IssuePageSize"] = math.Min(totalCount-count, maxGithubPageSize)
+		var respData model.Query
+		retryTimes := 10
+		for retryTimes != 0 {
+			err := request.QueryWithAuthPool(context.Background(), &respData, v)
+			if err != nil {
+				log.Errorf(err.Error()+" \n query variables: %v \n retry time: %d", v, 10-retryTimes)
```
Comment (mahjonp) [inline infrastructure/github/crawler/crawler.go:71]: ditto.

**Review by mahjonp — COMMENTED** [review]
I know a litter about this project, so give some comments on the code.

* Suggest using a full name as the go module name: https://golang.org/ref/mod#modules-overview
* Error subtype should be prefixed with Err
* Prefer to using a structured logger than `fmt.Printxxx` 

Rest LGTM, PTAL @sjwsl

File: infrastructure/github/crawler/crawler.go (hunk 3 above)
Comment (Andrewmatilde) [reply infrastructure/github/crawler/crawler.go:22]: Thanks for your review. I agree with you.

File: infrastructure/github/crawler/crawler.go (hunk 2 above)
Comment (Andrewmatilde) [reply infrastructure/github/crawler/crawler.go:8]: Thanks for your review. I agree with you.

File: infrastructure/github/crawler/client/client.go (hunk 1 above)
Comment (Andrewmatilde) [reply infrastructure/github/crawler/client/client.go:47]: Thanks for your review. I agree with you.

File: infrastructure/github/crawler/crawler.go (hunk 4 above)
Comment (Andrewmatilde) [reply infrastructure/github/crawler/crawler.go]: Thanks for your review. I agree with you.

**Comment by Andrewmatilde** [comment]
PTAL @sjwsl

File: infrastructure/github/crawler/client/client.go [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+var client *graphql.Client
+var clientIsOpen bool
+
+type Request struct {
+	req   *graphql.Request
+	index int
+}
+
+// NewClient return a client with request index 0
+func NewClient() Request {
```
Comment (zhouqiang-cl) [inline infrastructure/github/crawler/client/client.go:23]: Why always index 0? Could requests be empty?

File: infrastructure/github/crawler/config/config.go [hunk 7]
```diff
@@ -0,0 +1,40 @@
+package config
+
+import (
+	"encoding/json"
```
Comment (zhouqiang-cl) [inline infrastructure/github/crawler/config/config.go:4]: sort import

File: infrastructure/github/crawler/crawler.go [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+import (
+	"context"
+	"log"
+	"reflect"
+
+	"github.com/PingCAP-QE/dashboard/infrastructure/github/crawler/client"
+	"github.com/PingCAP-QE/dashboard/infrastructure/github/crawler/model"
+	"github.com/PingCAP-QE/dashboard/infrastructure/github/crawler/util"
+
+	"github.com/pkg/math"
```
Comment (zhouqiang-cl) [inline infrastructure/github/crawler/crawler.go:12]: sort import package

File: infrastructure/github/crawler/crawler.go [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+		"repo_name":       opt.RepoName,
+		"issueFilters":    opt.IssueFilters,
+		"issueAfter":      nil,
+		"commentAfter":    nil,
+		"tagAfter":        nil,
+		"IssuePageSize":   0,
+		"CommentPageSize": 0,
+		"tagPageSize":     0,
+	}
+	log.Printf("Ping count with %v\n", v)
```
Comment (zhouqiang-cl) [inline infrastructure/github/crawler/crawler.go:45]: Add a blank line?

File: infrastructure/github/crawler/crawler.go [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+func FetchByRepoSafe(request client.Request, opt FetchOption) *model.Query {
+	totalData := FetchByRepo(request, opt)
+	QueryCompletenessSpec(totalData)
+	QueryDataInvalidSpec(totalData)
+	return totalData
+}
+
+// FetchByRepo Fetch all the Query data necessary with FetchOption and request.
+func FetchByRepo(request client.Request, opt FetchOption) *model.Query {
+	v := map[string]interface{}{
```
Comment (zhouqiang-cl) [inline infrastructure/github/crawler/crawler.go:34]: I think we should make v more readable

File: infrastructure/github/crawler/crawler.go [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+				log.Fatal(err.Error()+" \n query variables: %v \n retry time: %d", v, 10-retryTimes)
+			}
+		}
+		log.Printf("Fetch success.>\n")
+		totalData.Repository.Refs.Nodes = append(totalData.Repository.Refs.Nodes, respData.Repository.Refs.Nodes...)
+		if !respData.Repository.Refs.PageInfo.HasNextPage {
+			break
+		}
+		v["tagAfter"] = respData.Repository.Refs.PageInfo.EndCursor
+	}
```
Comment (zhouqiang-cl) [inline infrastructure/github/crawler/crawler.go:119]: This function is too long, I think we should split it

File: infrastructure/github/crawler/model/query.go [hunk 12]
```diff
@@ -0,0 +1,89 @@
+package model
+
+import "time"
+
+type Query struct {
```
Comment (zhouqiang-cl) [inline infrastructure/github/crawler/model/query.go:5]: This struct is too big

File: infrastructure/github/crawler/client/client.go (hunk 6 above)
Comment (Andrewmatilde) [reply infrastructure/github/crawler/client/client.go:24]: Thanks for your review. I agree with you.

File: infrastructure/github/crawler/crawler.go (hunk 9 above)
Comment (Andrewmatilde) [reply infrastructure/github/crawler/crawler.go:46]: Thanks for your review. I agree with you.

File: infrastructure/github/crawler/model/query.go (hunk 12 above)
Comment (Andrewmatilde) [reply infrastructure/github/crawler/model/query.go:17]: Thanks for your review. I agree with you.

File: infrastructure/github/crawler/crawler.go (hunk 11 above)
Comment (Andrewmatilde) [reply infrastructure/github/crawler/crawler.go:120]: This function is just like the main function of crawler, not for scalability, I may write generator for it.

**Pull request merged**